lib/fs.js

```diff
@@ -827,117 +808,192 @@ fs.ftruncate = function(fd, len, callback) {
   binding.ftruncate(fd, len, req);
 };
 
 fs.ftruncateSync = function(fd, len) {
-  if (len === undefined) {
+  if (len == null)
     len = 0;
+  if (!Number.isInteger(fd) || fd < 0) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'fd',
+                               'unsigned integer');
+  }
+
+  if (!Number.isInteger(len) || len < 0) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'len',
+                               'unsigned integer');
   }
   return binding.ftruncate(fd, len);
 };
 
 fs.rmdir = function(path, callback) {
   callback = maybeCallback(callback);
-  if (handleError((path = getPathFromURL(path)), callback))
-    return;
-  if (!nullCheck(path, callback)) return;
-  var req = new FSReqWrap();
+  handleError(path = getPathFromURL(path));
+  nullCheck(path);
+
+  if (typeof path !== 'string' && !Buffer.isBuffer(path)) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'path',
+                               ['string', 'Buffer', 'URL']);
+  }
+
+  const req = new FSReqWrap();
   req.oncomplete = callback;
   binding.rmdir(pathModule._makeLong(path), req);
 };
 
 fs.rmdirSync = function(path) {
-  handleError((path = getPathFromURL(path)));
+  handleError(path = getPathFromURL(path));
   nullCheck(path);
+
+  if (typeof path !== 'string' && !Buffer.isBuffer(path)) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'path',
+                               ['string', 'Buffer', 'URL']);
+  }
+
   return binding.rmdir(pathModule._makeLong(path));
 };
 
 fs.fdatasync = function(fd, callback) {
-  var req = new FSReqWrap();
+  if (!Number.isInteger(fd) || fd < 0) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'fd',
+                               'unsigned integer');
+  }
+
+  const req = new FSReqWrap();
   req.oncomplete = makeCallback(callback);
   binding.fdatasync(fd, req);
 };
 
 fs.fdatasyncSync = function(fd) {
+  if (!Number.isInteger(fd) || fd < 0) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'fd',
+                               'unsigned integer');
+  }
+
   return binding.fdatasync(fd);
 };
 
 fs.fsync = function(fd, callback) {
-  var req = new FSReqWrap();
+  if (!Number.isInteger(fd) || fd < 0) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'fd',
+                               'unsigned integer');
+  }
+  const req = new FSReqWrap();
   req.oncomplete = makeCallback(callback);
   binding.fsync(fd, req);
 };
 
 fs.fsyncSync = function(fd) {
+  if (!Number.isInteger(fd) || fd < 0) {
+    throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'fd',
+                               'unsigned integer');
+  }
   return binding.fsync(fd);
 };
 
 fs.mkdir = function(path, mode, callback) {
   if (typeof mode === 'function') callback = mode;
   callback = makeCallback(callback);
-  if (handleError((path = getPathFromURL(path)), callback))
-    return;
-  if (!nullCheck(path, callback)) return;
-  var req = new FSReqWrap();
+  mode = modeNum(0o777);
```

Comment:
The modeNum wrapping here is obsolete. I think you actually wanted to do
```js
if (typeof mode === 'function') {
  callback = mode;
  mode = 0o777;
} else {
  mode = modeNum(mode, 0o777);
}
```

Reply:
The same applies to other functions.

Reply:
+1